File: core/config_manager.py
# -*- coding: utf-8 -*-
"""配置管理模块"""

import json
import os
import sys
import shutil

def get_base_dir():
    """获取基础目录（用户可写目录）
    
    - 打包后：返回exe所在目录
    - 开发环境：返回项目根目录
    """
    if getattr(sys, 'frozen', False):
        # 打包后：返回exe所在目录（用户可写）
        return os.path.dirname(sys.executable)
    else:
        # 开发环境路径
        return os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

def get_default_config_dir():
    """获取默认配置目录（打包在exe内部的只读配置）
    
    - 打包后：返回sys._MEIPASS（临时解压目录）
    - 开发环境：返回项目根目录
    """
    if getattr(sys, 'frozen', False):
        return sys._MEIPASS
    else:
        return os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

class ConfigManager:
    """管理治疗配置数据"""
    
    def __init__(self, config_path=None):
        if config_path is None:
            base_dir = get_base_dir()
            config_path = os.path.join(base_dir, 'data', 'treatment_config.json')
        self.config_path = config_path
        self.config = self._load_config()
    
    def _load_config(self):
        """加载配置文件
        
        优先从用户目录读取，如果没有则从exe内部复制默认配置
        """
        # 优先从用户可写目录读取
        if os.path.exists(self.config_path):
            with open(self.config_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        
        # 打包环境：从exe内部的默认配置复制
        if getattr(sys, 'frozen', False):
            default_path = os.path.join(get_default_config_dir(), 'data', 'treatment_config.json')
            if os.path.exists(default_path):
                os.makedirs(os.path.dirname(self.config_path), exist_ok=True)
                shutil.copy2(default_path, self.config_path)
                with open(self.config_path, 'r', encoding='utf-8') as f:
                    return json.load(f)
        
        # 都没有则返回默认空配置
        return {"hospital_name": "", "title": "治疗记录单", "treatments": []}
    
    def save_config(self):
        """保存配置文件"""
        os.makedirs(os.path.dirname(self.config_path), exist_ok=True)
        with open(self.config_path, 'w', encoding='utf-8') as f:
            json.dump(self.config, f, ensure_ascii=False, indent=2)
    
    def get_hospital_name(self):
        """获取医院名称"""
        return self.config.get('hospital_name', '')
    
    def set_hospital_name(self, name):
        """设置医院名称"""
        self.config['hospital_name'] = name
        self.save_config()
    
    def get_title(self):
        """获取标题"""
        return self.config.get('title', '治疗记录单')
    
    def set_title(self, title):
        """设置标题"""
        self.config['title'] = title
        self.save_config()
    
    def get_treatments(self):
        """获取所有治疗项目"""
        return self.config.get('treatments', [])
    
    def get_treatment_by_id(self, treatment_id):
        """根据ID获取治疗项目"""
        for treatment in self.get_treatments():
            if treatment['id'] == treatment_id:
                return treatment
        return None
    
    def get_treatment_names(self):
        """获取所有治疗项目名称列表"""
        return [t['name'] for t in self.get_treatments()]
    
    def get_treatment_duration(self, treatment_id):
        """获取治疗项目时长"""
        treatment = self.get_treatment_by_id(treatment_id)
        if treatment:
            return treatment.get('duration', '')
        return ''
    
    def get_diagnoses_by_treatment(self, treatment_id):
        """根据治疗项目ID获取诊断列表"""
        treatment = self.get_treatment_by_id(treatment_id)
        if treatment:
            return treatment.get('diagnoses', [])
        return []
    
    def get_treatment_details(self, treatment_id, diagnosis_id):
        """获取治疗内容"""
        diagnoses = self.get_diagnoses_by_treatment(treatment_id)
        for diag in diagnoses:
            if diag['id'] == diagnosis_id:
                return diag.get('details', '')
        return ''
    
    def add_treatment(self, name, duration=''):
        """添加治疗项目"""
        treatments = self.get_treatments()
        # 生成新ID
        max_id = 0
        for t in treatments:
            if t['id'].startswith('t'):
                try:
                    num = int(t['id'][1:])
                    if num > max_id:
                        max_id = num
                except ValueError:
                    pass
        new_id = f't{max_id + 1}'
        new_treatment = {
            'id': new_id,
            'name': name,
            'duration': duration,
            'diagnoses': []
        }
        self.config['treatments'].append(new_treatment)
        self.save_config()
        return new_treatment
    
    def update_treatment(self, treatment_id, name, duration=''):
        """更新治疗项目名称和时长"""
        for treatment in self.config['treatments']:
            if treatment['id'] == treatment_id:
                treatment['name'] = name
                treatment['duration'] = duration
                self.save_config()
                return True
        return False
    
    def delete_treatment(self, treatment_id):
        """删除治疗项目"""
        self.config['treatments'] = [
            t for t in self.config['treatments'] if t['id'] != treatment_id
        ]
        self.save_config()
    
    def add_diagnosis(self, treatment_id, name, details=''):
        """添加诊断"""
        treatment = self.get_treatment_by_id(treatment_id)
        if treatment:
            # 生成新ID
            max_id = 0
            for d in treatment.get('diagnoses', []):
                if d['id'].startswith('d'):
                    try:
                        num = int(d['id'][1:])
                        if num > max_id:
                            max_id = num
                    except ValueError:
                        pass
            new_id = f'd{max_id + 1}'
            new_diagnosis = {
                'id': new_id,
                'name': name,
                'details': details
            }
            if 'diagnoses' not in treatment:
                treatment['diagnoses'] = []
            treatment['diagnoses'].append(new_diagnosis)
            self.save_config()
            return new_diagnosis
        return None
    
    def update_diagnosis(self, treatment_id, diagnosis_id, name, details):
        """更新诊断"""
        treatment = self.get_treatment_by_id(treatment_id)
        if treatment:
            for diag in treatment.get('diagnoses', []):
                if diag['id'] == diagnosis_id:
                    diag['name'] = name
                    diag['details'] = details
                    self.save_config()
                    return True
        return False
    
    def delete_diagnosis(self, treatment_id, diagnosis_id):
        """删除诊断"""
        treatment = self.get_treatment_by_id(treatment_id)
        if treatment:
            treatment['diagnoses'] = [
                d for d in treatment.get('diagnoses', []) if d['id'] != diagnosis_id
            ]
            self.save_config()
    
    def get_surcharges(self):
        """获取所有加收配置"""
        return self.config.get('surcharges', [])
    
    def get_surcharge_titles(self):
        """获取加收职称列表"""
        return [s['title'] for s in self.get_surcharges()]
    
    def get_surcharge_items_by_title(self, title):
        """根据职称获取加收项目列表"""
        for surcharge in self.get_surcharges():
            if surcharge['title'] == title:
                return surcharge.get('items', [])
        return []
    
    def is_acupuncture_treatment(self, treatment_name):
        """判断是否为针灸类治疗项目"""
        return '针灸' in treatment_name

File: TreatmentHelper/Services/ConfigManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using TreatmentHelper.Models;

namespace TreatmentHelper.Services;

/// <summary>
/// 配置管理服务
/// </summary>
public class ConfigManager
{
    private readonly string _configPath;
    private TreatmentConfig _config = new();

    public ConfigManager()
    {
        var baseDir = AppDomain.CurrentDomain.BaseDirectory;
        _configPath = Path.Combine(baseDir, "Config", "treatment_config.json");
        LoadConfig();
    }

    public ConfigManager(string configPath)
    {
        _configPath = configPath;
        LoadConfig();
    }

    private void LoadConfig()
    {
        try
        {
            if (File.Exists(_configPath))
            {
                var json = File.ReadAllText(_configPath);
                _config = JsonSerializer.Deserialize<TreatmentConfig>(json) ?? new TreatmentConfig();
            }
            else
            {
                // 创建默认配置目录
                Directory.CreateDirectory(Path.GetDirectoryName(_configPath)!);
                _config = GetDefaultConfig();
                SaveConfig();
            }
        }
        catch
        {
            _config = new TreatmentConfig();
        }
    }

    private TreatmentConfig GetDefaultConfig()
    {
        return new TreatmentConfig
        {
            HospitalName = "",
            Title = "治疗记录单",
            Surcharges = new List<Surcharge>
            {
                new() { Title = "主任医师", Items = new List<string> { "常规针法加收×1", "特殊手法针法加收×1" } },
                new() { Title = "副主任医师", Items = new List<string> { "常规针法加收×1", "特殊手法针法加收×1" } }
            },
            Treatments = new List<Treatment>()
        };
    }

    public void SaveConfig()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(_configPath)!);
        var options = new JsonSerializerOptions
        {
            WriteIndented = true,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
        };
        var json = JsonSerializer.Serialize(_config, options);
        File.WriteAllText(_configPath, json);
    }

    // 医院名称
    public string GetHospitalName() => _config.HospitalName;
    public void SetHospitalName(string name)
    {
        _config.HospitalName = name;
        SaveConfig();
    }

    // 标题
    public string GetTitle() => _config.Title;
    public void SetTitle(string title)
    {
        _config.Title = title;
        SaveConfig();
    }

    // 治疗项目
    public List<Treatment> GetTreatments() => _config.Treatments;

    public Treatment? GetTreatmentById(string id) =>
        _config.Treatments.FirstOrDefault(t => t.Id == id);

    public List<string> GetTreatmentNames() =>
        _config.Treatments.Select(t => t.Name).ToList();

    public string GetTreatmentDuration(string treatmentId)
    {
        var treatment = GetTreatmentById(treatmentId);
        return treatment?.Duration ?? "";
    }

    // 诊断
    public List<Diagnosis> GetDiagnosesByTreatment(string treatmentId)
    {
        var treatment = GetTreatmentById(treatmentId);
        return treatment?.Diagnoses ?? new List<Diagnosis>();
    }

    public string GetTreatmentDetails(string treatmentId, string diagnosisId)
    {
        var diagnoses = GetDiagnosesByTreatment(treatmentId);
        var diag = diagnoses.FirstOrDefault(d => d.Id == diagnosisId);
        return diag?.Details ?? "";
    }

    // 添加治疗项目
    public Treatment AddTreatment(string name, string duration = "")
    {
        var maxId = _config.Treatments
            .Where(t => t.Id.StartsWith("t"))
            .Select(t => int.TryParse(t.Id[1..], out var num) ? num : 0)
            .DefaultIfEmpty(0)
            .Max();

        var newTreatment = new Treatment
        {
            Id = $"t{maxId + 1}",
            Name = name,
            Duration = duration,
            Diagnoses = new List<Diagnosis>()
        };

        _config.Treatments.Add(newTreatment);
        SaveConfig();
        return newTreatment;
    }

    // 更新治疗项目
    public bool UpdateTreatment(string treatmentId, string name, string duration = "")
    {
        var treatment = GetTreatmentById(treatmentId);
        if (treatment == null) return false;

        treatment.Name = name;
        treatment.Duration = duration;
        SaveConfig();
        return true;
    }

    // 删除治疗项目
    public void DeleteTreatment(string treatmentId)
    {
        _config.Treatments.RemoveAll(t => t.Id == treatmentId);
        SaveConfig();
    }

    // 添加诊断
    public Diagnosis? AddDiagnosis(string treatmentId, string name, string details = "")
    {
        var treatment = GetTreatmentById(treatmentId);
        if (treatment == null) return null;

        var maxId = treatment.Diagnoses
            .Where(d => d.Id.StartsWith("d"))
            .Select(d => int.TryParse(d.Id[1..], out var num) ? num : 0)
            .DefaultIfEmpty(0)
            .Max();

        var newDiagnosis = new Diagnosis
        {
            Id = $"d{maxId + 1}",
            Name = name,
            Details = details
        };

        treatment.Diagnoses.Add(newDiagnosis);
        SaveConfig();
        return newDiagnosis;
    }

    // 更新诊断
    public bool UpdateDiagnosis(string treatmentId, string diagnosisId, string name, string details)
    {
        var treatment = GetTreatmentById(treatmentId);
        if (treatment == null) return false;

        var diagnosis = treatment.Diagnoses.FirstOrDefault(d => d.Id == diagnosisId);
        if (diagnosis == null) return false;

        diagnosis.Name = name;
        diagnosis.Details = details;
        SaveConfig();
        return true;
    }

    // 删除诊断
    public void DeleteDiagnosis(string treatmentId, string diagnosisId)
    {
        var treatment = GetTreatmentById(treatmentId);
        if (treatment == null) return;

        treatment.Diagnoses.RemoveAll(d => d.Id == diagnosisId);
        SaveConfig();
    }

    // 加收配置
    public List<Surcharge> GetSurcharges() => _config.Surcharges;

    public List<string> GetSurchargeTitles() =>
        _config.Surcharges.Select(s => s.Title).ToList();

    public List<string> GetSurchargeItemsByTitle(string title)
    {
        var surcharge = _config.Surcharges.FirstOrDefault(s => s.Title == title);
        return surcharge?.Items ?? new List<string>();
    }

    // 判断是否为针灸类治疗
    public bool IsAcupunctureTreatment(string treatmentName) =>
        treatmentName.Contains("针灸");
}
